File: kotlin-node/karakum/src/jsMain/resources/plugins/convertImplicitArrayBuffer.js
import ts from "typescript";
import * as karakum from "../karakum.mjs";

export default function (node, context, render) {
    const sourceFileName = node.getSourceFile()?.fileName ?? "generated.d.ts"

    if (
        sourceFileName.endsWith("buffer.buffer.d.ts")
        && ts.isConditionalTypeNode(node)
        && node.parent
        && ts.isTypeAliasDeclaration(node.parent)
        && node.parent.name.text === "ImplicitArrayBuffer"
    ) {
        const typeScriptService = context.lookupService(karakum.typeScriptServiceKey.get())

        return `js.buffer.ArrayBufferLike /* ${typeScriptService?.printNode(node)} */`;
    }

    return null
}

File: kotlin-node/karakum/src/jsMain/kotlin/node/karakum/plugins/convertImplicitArrayBuffer.kt
package node.karakum.plugins

import node.karakum.util.nullable
import io.github.sgrishchenko.karakum.extension.createPlugin
import io.github.sgrishchenko.karakum.extension.plugins.TypeScriptService
import io.github.sgrishchenko.karakum.extension.plugins.typeScriptServiceKey
import io.github.sgrishchenko.karakum.util.getParentOrNull
import io.github.sgrishchenko.karakum.util.getSourceFileOrNull
import typescript.isConditionalTypeNode
import typescript.isTypeAliasDeclaration

val convertImplicitArrayBuffer = createPlugin { node, context, _ ->
    nullable {
        val sourceFileName = ensureNotNull(node.getSourceFileOrNull()).fileName
        ensure(sourceFileName.endsWith("buffer.buffer.d.ts"))

        ensure(isConditionalTypeNode(node))

        val typeAlias = ensureNotNull(node.getParentOrNull())
        ensure(isTypeAliasDeclaration(typeAlias))
        ensure(typeAlias.name.text == "ImplicitArrayBuffer")

        val typeScriptService = ensureNotNull(context.lookupService<TypeScriptService>(typeScriptServiceKey))

        "js.buffer.ArrayBufferLike /* ${typeScriptService.printNode(node)} */"
    }
}
